simSPI/helpers.py

```diff
@@ -0,0 +1,308 @@
+"""Helper functions for tem.py."""
+import math
+import os
+
+import h5py
+import mdtraj as md
+import mrcfile
+import numpy as np
+from scipy.spatial.distance import pdist
+from scipy.stats import special_ortho_group
+
+
+def mrc2data(mrc_file=None):
+    """Return micrograph from an input .mrc file.
+
+    Parameters
+    ----------
+    mrc_file : str
+        File name for .mrc file to turn into micrograph
+    """
+    if mrc_file is not None:
+        with mrcfile.open(mrc_file, "r", permissive=True) as mrc:
+            micrograph = mrc.data
+        if micrograph is not None:
+            if len(micrograph.shape) == 2:
+                micrograph = micrograph[np.newaxis, ...]
+        else:
+            print("Warning! Data in {} is None...".format(mrc_file))
+        return micrograph
+    return None
+
+
+def data_and_dic_2hdf5(data, h5_file, dic=None):
+    """Save a dictionary which might contain different data types to a .hdf5 file.
+
+    Parameters
+    ----------
+    data : Object
+        Data to store.
+    h5_file : str
+        Relative path to write .h5 file to.
+    dic : dict
+        Dictionary containing fields to save.
+    """
+    if dic is None:
+        dic = {}
+    dic["data"] = data
+    with h5py.File(h5_file, "w") as file:
+        recursively_save_dict_contents_to_group(file, "/", dic)
+
+
+def recursively_save_dict_contents_to_group(h5file, path, dic):
+    """Recursively save dictionary contents to group.
+
+    Parameters
+    ----------
+    h5file : File
+        .hdf5 file to write to.
+    path : str
+        Relative path to save dictionary contents.
+    dic : dict
+        Dictionary containing data.
+    """
+    for k, v in dic.items():
+        if isinstance(v, (np.ndarray, np.int64, np.float64, int, float, str, bytes)):
+            h5file[path + k] = v
+        elif isinstance(v, type(None)):
+            h5file[path + k] = str("None")
+        elif isinstance(v, dict):
+            recursively_save_dict_contents_to_group(h5file, path + k + "/", v)
+        else:
+            raise ValueError("Cannot save %s type" % type(v))
```

Comment:
I like this type of defensive coding, so we know when and why things fail. I had to move `assert False` to tests because no asserts were allowed in non-tests. I didn't realize that `raise valueError` was allowed (hopefully it really is).